docs/apis/plugintypes/repository/index.mdx

```diff
@@ -0,0 +1,975 @@
+<details>
+  <summary>View example</summary>
+  <div>
+
+```php
+/**
+ * Get file listing.
+ *
+ * This is a mandatory method for any repository.
+ *
+ * See repository::get_listing() for details.
+ *
+ * @param string $encodedpath
+ * @param string $page
+ * @return array the list of files, including meta infomation
+ */
+public function get_listing($encodedpath = '', $page = '') {
+    // This methods
+    return [
+        //this will be used to build navigation bar.
+        'path'=>[
+            [
+                'name'=>'root'
+                'path'=>'/'
+            ],
+            [
+                'name'=>'subfolder',
+                'path'=>'/subfolder'
+            ],
+        ],
+        'manage'=>'http://webmgr.moodle.com',
+        'list'=> [
+            [
+                'title'=>'filename1',
+                'date'=>'1340002147',
+                'size'=>'10451213',
+                'source'=>'http://www.moodle.com/dl.rar',
+            ],
+            [
+                'title'=>'folder',
+                'date'=>'1340002147',
+                'size'=>'0',
+                'children'=>[],
+            ],
+        ],
+    ];
+}
+```
+
+  </div>
+</details>
+
+Amongst other details, this returns a **title** for each file (to be displayed in the filepicker) and the **source** for the file (which will be included in the request to 'download' the file into Moodle or to generate a link to the file). Directories return a **children** value, which is either an empty array (if 'dynload' is specified) or an array of the files and directories contained within it.
+
+<details>
+  <summary>The full specification of list element</summary>
+  <div>
+
+import FullList from '!!raw-loader!./_examples/full_list.php';
+
+<CodeBlock language="php">{FullList}</CodeBlock>
+
+  </div>
+</details>
+
+### Dynamically loading
+
+Some repositories contain many files which cannot load in one time, in this case, we need dynamically loading to fetch them step by step, files in subfolder won't be listed until user click the folder in file picker treeview.
+
+As a plug-in developer, if you set dynload flag as **true**, you should return files and folders (set children as a null array) in current path only instead of building the whole file tree.
+
+The use of the **object** tag, instead of returning a **list** of files, allows you to embed an external file chooser within the repository panel. See [[Repository plugins embedding external file chooser]] for details about how to do this.
+
+### User login (optional)
+
+If the plugin requires login from the user at the time when they use it, then these functions can be used.
+
+#### print_login
+
+Returns an array of the elements required in the login form. If no login form is required, then the default implementation of this will redirect to the files list. If $this->options[is not set, then an HTML-snippet with the login fields (but not the form tags) should be output, instead of returning the form details.
```

Comment:
Do we really need this level of detail? It *should* be in the phpdoc.

We shouldn't aim to replace or duplicate it all.

This stanza does not make sense - it looks like part of the sentence is missing (`If $this->options[is not se, then ,...`).

```suggestion
For plugins which need to support login to a remote service, the `print_login()` function can be used to return an array of the form elements needed to support the login.

... (not sure what the rest of the intent of this sentence was to complete it)
```

Reply:
Ok review this point and clarify the content. I also moved the ajax part to a note.

I noticed that print_login method returns an array or an HTML form depending on the $this->options['ajax'] value. I wouldn't say I like it and probably we should stop doing that but, for now, it should be documented as it is.